components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/utils"
+)
+
+const model = "rtsp"
+
+func init() {
+	registry.RegisterComponent(camera.Subtype, model, registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+			attrs, ok := cfg.ConvertedAttributes.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(attrs, cfg.ConvertedAttributes)
+			}
+			return NewRTSPCamera(ctx, attrs, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(
+		camera.SubtypeName,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Attrs
+			attrs, err := config.TransformAttributeMapToStruct(&conf, attributes)
+			if err != nil {
+				return nil, err
+			}
+			result, ok := attrs.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(result, attrs)
+			}
+			return result, nil
+		},
+		&Attrs{},
+	)
+}
+
+// Attrs are the config attributes for an RTSP camera model.
+type Attrs struct {
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	client     *gortsplib.Client
+	decoder    *h264Decoder
+	cancelFunc context.CancelFunc
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	clientTerminated := liberrors.ErrClientTerminated{}
+	rc.cancelFunc()
+	err := rc.client.Close()
+	if err != nil && !errors.Is(err, clientTerminated) {
+		return err
+	}
+	return rc.decoder.Close()
+}
+
+// NewRTSPCamera creates a camera client for an RTSP given given the server URL.
+// Right now, only supports servers that have H264 video tracks.
+func NewRTSPCamera(ctx context.Context, attrs *Attrs, logger golog.Logger) (camera.Camera, error) {
+	c := &gortsplib.Client{}
+	// parse URL
+	u, err := url.Parse(attrs.Address)
+	if err != nil {
+		return nil, err
+	}
+	// connect to the server - be sure to close it if setup fails.
+	err = c.Start(u.Scheme, u.Host)
+	if err != nil {
+		return nil, err
+	}
+	// find published tracks
+	tracks, baseURL, _, err := c.Describe(u)
+	if err != nil {
+		return nil, multierr.Combine(err, c.Close())
+	}
+
+	// find the H264 track
+	track := func() *gortsplib.TrackH264 {
+		for _, track := range tracks {
+			if track, ok := track.(*gortsplib.TrackH264); ok {
+				return track
+			}
+		}
+		return nil
+	}()
+	if track == nil {
+		err := errors.Errorf("H264 track not found in rtsp camera %s", u)
+		return nil, multierr.Combine(err, c.Close())
+	}
```

Comment:
Note: Currently only H264 decoding is available. Creation of the camera will fail if a H264 track is not available.